File: Data Analyst in R/Step 5 - Probability and Statistics/2. Statistics Intermediate in R Averages and Variability/4. Measures of Variability.r
## 1. The Range ##

library(readr)
# Heads up that the following instruction can generate a column specification warnings, you can ignore that.
# These warnings are related to that `readr` has misguidedly guessed the type of values in the column `Pool QC`. 
# To avoid this warning we can manually specify the type of this column the `col_types` parameter.
houses  <-  read_tsv('AmesHousing_1.txt', col_types = cols(`Pool QC` = col_character()))
library(dplyr)

sale_price_range <- houses %>%
    group_by(`Yr Sold`) %>%
    summarize(range_by_year = max(SalePrice) - min(SalePrice))

one  <-  FALSE
two  <-  TRUE

## 2. The Average Distance ##

C  <-  c(1,1,1,1,1,1,1,1,1,21)
average_distance <- function(vector) {
    distances  <-  vector - mean(vector)
    sum(distances) / length(distances)
}

avg_distance  <-  average_distance(C)

## 3. Mean Absolute Deviation ##

C  <-  c(1,1,1,1,1,1,1,1,1,21)

#average_distance <- function(vector) {
#    distances  <-  vector - mean(vector)
#    sum(distances) / length(distances)
#}
mean_absolute_deviation <- function(vector) {
    distances  <-  abs(vector - mean(vector)) #we only add abs function here
    sum(distances) / length(distances)
}

mad  <-  mean_absolute_deviation(C)

## 4. Variance ##

C  <-  c(1,1,1,1,1,1,1,1,1,21)

#mean_absolute_deviation <- function(vector) {
#    distances  <-  abs(vector - mean(vector))
#    sum(distances) / length(distances)
#}
variance <- function(vector) {
    distances  <-  (vector - mean(vector))**2 #we only need to compute the squared distances here
    sum(distances) / length(distances)
}

variance_C  <-  variance(C)

## 5. Standard Deviation ##

C  <-  c(1,1,1,1,1,1,1,1,1,21)

#variance <- function(vector) {
#    distances  <-  (vector - mean(vector))**2 #we only need to compute the squared distances here
#    sum(distances) / length(distances)
#}
standard_deviation <- function(vector) {
    distances  <-  (vector - mean(vector))**2 
    sqrt(sum(distances) / length(distances)) #adding sqrt at this line
}

standard_deviation_C  <-  standard_deviation(C)

## 6. Average Variability Around the Mean ##

library(dplyr)

standard_deviation <- function(vector) {
    distances  <-  (vector - mean(vector))**2 #we only need to compute the squared distances here
    sqrt(sum(distances) / length(distances))
}
# Measure first the variability for each year
houses_years_std <- houses %>%
    group_by(`Yr Sold`) %>%
    summarize(st_dev = standard_deviation(SalePrice)) %>%
    arrange(st_dev)

# Get years of max and min variability
greatest_variability  <-  houses_years_std %>%
  filter(st_dev == max(st_dev)) %>% 
  pull(`Yr Sold`)

lowest_variability  <-  houses_years_std %>%
  filter(st_dev == min(st_dev)) %>% 
  pull(`Yr Sold`)

## 7. A Measure of Spread ##

set.seed(10)

sample1  <-  sample(x = houses$`Year Built`, size = 50)
sample2  <-  sample(x = houses$`Year Built`, size = 50)

standard_deviation <- function(vector) {
    distances  <-  (vector - mean(vector))**2 #we only need to compute the squared distances here
    sqrt(sum(distances) / length(distances))
}
bigger_spread  <-  'sample 2'
st_dev1  <-  standard_deviation(sample1)
st_dev2  <-  standard_deviation(sample2)

## 8. The Sample Standard Deviation ##

standard_deviation <- function(vector) {
    distances  <-  (vector - mean(vector))**2 #we only need to compute the squared distances here
    sqrt(sum(distances) / length(distances))
}

library(ggplot2)
set.seed(1)
std_points  <-  replicate(n = 5000, expr = standard_deviation(sample(x = houses$SalePrice, size = 10)))

std_points_tibble <- tibble::tibble(std_points)

ggplot(data = std_points_tibble, aes(x = std_points)) +
    geom_histogram(bins = 10, position = "identity") +
    geom_vline(aes(xintercept = standard_deviation(houses$SalePrice))) +
    xlab("Sample standard deviation") + 
    ylab("Frequency")

## 10. Bessel's Correction ##

standard_deviation <- function(vector) {
    distances  <-  (vector - mean(vector))**2 #we only need to compute the squared distances here
    sqrt(sum(distances) / length(distances))
}

library(ggplot2)
set.seed(1)

std_points  <-  replicate(n = 5000, expr = standard_deviation(sample(x = houses$SalePrice, size = 10)))

#ggplot(data = tibble::tibble(std_points), aes(x = std_points)) +
#    geom_histogram(bins = 10, position = "identity") +
#    geom_vline(aes(xintercept = standard_deviation(houses$SalePrice))) +
#    xlab("Sample standard deviation") + 
#    ylab("Frequency")

population_stdev  <-  standard_deviation(houses$SalePrice)
standard_deviation_bessel_correction <- function(vector) {
    distances  <-  (vector - mean(vector))**2 
    sqrt(sum(distances) / (length(distances) - 1) ) #Only this line has to change
}

library(ggplot2)
set.seed(1)

std_points  <-  replicate(n = 5000, expr = standard_deviation_bessel_correction(sample(x = houses$SalePrice, size = 10)))

std_points_tibble <- tibble::tibble(std_points)

ggplot(data = std_points_tibble, aes(x = std_points)) +
    geom_histogram(bins = 10, position = "identity") +
    geom_vline(aes(xintercept = population_stdev)) +
    xlab("Sample standard deviation") + 
    ylab("Frequency")

## 11. Standard Notation ##

set.seed(1)

sample_sales  <-  sample(x=houses$SalePrice, size=100)

variance_bessel_correction <- function(vector) {
    distances  <-  (vector - mean(vector))**2 #we only need to compute the squared distances here
    sum(distances) / (length(distances) - 1)
}

standard_deviation_bessel_correction <- function(vector) {
    distances  <-  (vector - mean(vector))**2 #we only need to compute the squared distances here
    sqrt(sum(distances) / (length(distances)-1) )
}
computed_stdev  <-  standard_deviation_bessel_correction(sample_sales)
stdev_r  <-  sd(sample_sales)
equal_stdevs  <-  computed_stdev == stdev_r

computed_var  <-  variance_bessel_correction(sample_sales)
var_r  <-  var(sample_sales)
equal_vars  <-  computed_var == var_r

## 12. Sample Variance — Unbiased Estimator ##

population  <-  c(0, 3, 6)

samples  <-  list(c(0,3), c(0,6),
               c(3,0), c(3,6),
               c(6,0), c(6,3)
)

variance <- function(vector) {
    distances  <-  (vector - mean(vector))**2 #we only need to compute the squared distances here
    sum(distances) / length(distances)
}

standard_deviation <- function(vector) {
    distances  <-  (vector - mean(vector))**2 #we only need to compute the squared distances here
    sqrt(sum(distances) / length(distances) )
}
population_var  <-  variance(population)
population_std  <-  standard_deviation(population)

st_devs  <-  purrr::map_dbl(samples,sd)
variances  <-  purrr::map_dbl(samples,var)

mean_std  <-  mean(st_devs)
mean_var  <-  mean(variances)

equal_stdev  <-  population_std == mean_std
equal_var  <-  population_var == mean_var

File: Data Analyst in Python/Step 5 - Probability and Statistics/2. Statistics Intermediate Averages and Variability/4. Measures of Variability.py
## 1. The Range ##

import pandas as pd
houses = pd.read_table('AmesHousing_1.txt')
def find_range(array):
    return max(array) - min(array)

range_by_year = {}
for year in houses['Yr Sold'].unique():
    data_by_year = houses[houses['Yr Sold'] == year]
    range_by_year[year] = find_range(data_by_year['SalePrice'])
    
one = False
two = True

## 2. The Average Distance ##

C = [1,1,1,1,1,1,1,1,1,21]
def average_distance(array):
    reference_point = sum(array) / len(array)
    
    distances = []
    for value in array:
        distances.append(value - reference_point)
        
    return sum(distances) / len(distances)

avg_distance = average_distance(C)

## 3. Mean Absolute Deviation ##

C = [1,1,1,1,1,1,1,1,1,21]
def mean_absolute_deviation(array):
    reference_point = sum(array) / len(array)
    
    distances = []
    for value in array:
        absolute_distance = abs(value - reference_point)
        distances.append(absolute_distance)
        
    return sum(distances) / len(distances)

mad = mean_absolute_deviation(C)

## 4. Variance ##

C = [1,1,1,1,1,1,1,1,1,21]
def variance(array):
    reference_point = sum(array) / len(array)
    
    distances = []
    for value in array:
        squared_distance = (value - reference_point)**2
        distances.append(squared_distance)
        
    return sum(distances) / len(distances)

variance_C = variance(C)

## 5. Standard Deviation ##

from math import sqrt
C = [1,1,1,1,1,1,1,1,1,21]
def standard_deviation(array):
    reference_point = sum(array) / len(array)
    
    distances = []
    for value in array:
        squared_distance = (value - reference_point)**2
        distances.append(squared_distance)
        
    variance = sum(distances) / len(distances)
    
    return sqrt(variance)

standard_deviation_C = standard_deviation(C)

## 6. Average Variability Around the Mean ##

def standard_deviation(array):
    reference_point = sum(array) / len(array)
    
    distances = []
    for value in array:
        squared_distance = (value - reference_point)**2
        distances.append(squared_distance)
        
    variance = sum(distances) / len(distances)
    
    return sqrt(variance)
# Measure first the variability for each year
years = {}

for year in houses['Yr Sold'].unique():
    year_segment = houses[houses['Yr Sold'] == year]
    st_dev = standard_deviation(year_segment['SalePrice'])
    years[year] = st_dev

# Get years of max and min variability
greatest_variability = max(years, key = years.get) # outputs directly the year with the maximum variability
lowest_variability = min(years, key = years.get) # outputs directly the year with the minimum variability

## 7. A Measure of Spread ##

sample1 = houses['Year Built'].sample(50, random_state = 1)
sample2 = houses['Year Built'].sample(50, random_state = 2)

def standard_deviation(array):
    reference_point = sum(array) / len(array)
    
    distances = []
    for value in array:
        squared_distance = (value - reference_point)**2
        distances.append(squared_distance)
    
    variance = sum(distances) / len(distances)
    
    return sqrt(variance)
bigger_spread = 'sample 2'
st_dev1 = standard_deviation(sample1)
st_dev2 = standard_deviation(sample2)

## 8. The Sample Standard Deviation ##

def standard_deviation(array):
    reference_point = sum(array) / len(array)
    
    distances = []
    for value in array:
        squared_distance = (value - reference_point)**2
        distances.append(squared_distance)
    
    variance = sum(distances) / len(distances)
    
    return sqrt(variance)

import matplotlib.pyplot as plt
st_devs = []

for i in range(5000):
    sample = houses['SalePrice'].sample(10, random_state = i)
    st_dev = standard_deviation(sample)
    st_devs.append(st_dev)
    
plt.hist(st_devs)
plt.axvline(standard_deviation(houses['SalePrice']))

## 9. Bessel's Correction ##

from math import sqrt
def standard_deviation(array):
    reference_point = sum(array) / len(array)
    
    distances = []
    for value in array:
        squared_distance = (value - reference_point)**2
        distances.append(squared_distance)
    
    variance = sum(distances) / len(distances)
    
    return sqrt(variance)

import matplotlib.pyplot as plt
st_devs = []

for i in range(5000):
    sample = houses['SalePrice'].sample(10, random_state = i)
    st_dev = standard_deviation(sample)
    st_devs.append(st_dev)

#plt.hist(st_devs)
#plt.axvline(pop_stdev)  # pop_stdev is pre-saved from the last screen
from math import sqrt

def standard_deviation(array):
    reference_point = sum(array) / len(array)
    
    distances = []
    for value in array:
        squared_distance = (value - reference_point)**2
        distances.append(squared_distance)
    
    variance = sum(distances) / (len(distances) - 1)
    
    return sqrt(variance)

import matplotlib.pyplot as plt
st_devs = []

for i in range(5000):
    sample = houses['SalePrice'].sample(10, random_state = i)
    st_dev = standard_deviation(sample)
    st_devs.append(st_dev)
    
plt.hist(st_devs)
plt.axvline(pop_stdev)

## 10. Standard Notation ##

sample = houses.sample(100, random_state = 1)
from numpy import std, var
pandas_stdev = sample['SalePrice'].std(ddof = 1) # default ddof = 1
numpy_stdev = std(sample['SalePrice'], ddof = 1) # default ddof = 0
equal_stdevs = pandas_stdev == numpy_stdev

pandas_var = sample['SalePrice'].var(ddof = 1) # default ddof = 1
numpy_var = var(sample['SalePrice'], ddof = 1) #default ddof = 0
equal_vars = pandas_var == numpy_var

## 11. Sample Variance — Unbiased Estimator ##

population = [0, 3, 6]

samples = [[0,3], [0,6],
           [3,0], [3,6],
           [6,0], [6,3]
          ]
from numpy import var, std

pop_var = var(population, ddof = 0)
pop_std = std(population, ddof = 0)

st_devs = []
variances = []

for sample in samples:
    st_devs.append(std(sample, ddof = 1))
    variances.append(var(sample, ddof = 1))
    
mean_std = sum(st_devs) / len(st_devs)
mean_var = sum(variances) / len(variances)

equal_stdev = pop_std == mean_std
equal_var = pop_var == mean_var